code-dot-org/code-dot-org
# Webpack hot module replacement (HMR)

# Enable React "Fast Refresh" aka HMR
Reloads react components as soon as they build, without having to manually reload the page, works with both JSX and TSX.

## Demo
https://github.com/code-dot-org/code-dot-org/assets/223277/8beebf42-ca4c-487a-83d8-8b2876507f04

## What's Changed
1. Uses [react-refresh](https://www.npmjs.com/package/react-refresh), [@pmmmwh/react-refresh-webpack-plugin](https://github.com/pmmmwh/react-refresh-webpack-plugin), and [webpack-dev-server](https://www.npmjs.com/package/webpack-dev-server)
2. `yarn start` now launches [webpack-dev-server](https://webpack.js.org/configuration/dev-server/), which serves JS requests from memory and includes hooks to update the browser when source changes. It additionally proxies requests from port 9000 to port 3000 to provide a full dashboard server.
3. Prints new Dashboard URL at the end of `yarn start`: to use HMR, Dashboard must be accessed on port 9000: http://localhost-studio.code.org:9000. 
    <img width="601" alt="image" src="https://github.com/code-dot-org/code-dot-org/assets/223277/1b6210ea-cdb0-489b-a882-584beec434ad">
4. Removes createWebpackConfig() watch argument/mode, replace with `writeToDisk: true`, so those wishing to bypass HMR can still access pages directly on http://localhost-studio.code.org:3000. Also removes WebpackNotifierPlugin, as watch mode is now gone (and didn't work anyway).
5. Removes LiveReloadPlugin and envConstants.AUTO_RELOAD, which didn't seem to be working anyway, and are outdated approaches compared to HMR.
6. Removes `react-hot-loader`, another dependency previously used for hot reloading, but also wasn't working currently and has been deprecated by the maintainers.
7. Supports TSX by adding react-refresh-typescript plugin to ts-loader
8. Supports JSX by adding react-refresh/babel plugin to babel-loader
9. Renames `cdoBlockStyles.mjs` to `cdoBlockStyles.js`, converts to CJS style module, and updates references. This was required because `"type": "module"` (which is implied by the .mjs prefix) in the browser triggers issues with `react-refresh`. Since we only include this one .mjs file in the browser (and here only because the file is shared between node and browser), it was easy to rename to workaround.
10. Moves babel-cache from .babel-cache into the build dir, now is: build/babel-cache, so it gets cleared by yarn clean, etc
11. Simplify `createWebpackConfig()` calls using default values, emphasizes what makes targets different in Gruntfile.js
12. When running `grunt dev`, sets HOT=1 and DEV=1 in Gruntfile.js, removing repetition from package.json
13. Creates a `yarn grunt` target that sets `NODE_OPTIONS=--max-old-space-size=4096`, and use to remove 5x repetition from package.json, as well as no longer requiring explicit paths into node_modules, e.g.
    - Before: `"start": "DEV=1 node --max_old_space_size=4096 ./node_modules/.bin/grunt dev;"`
    - After: `"start": "yarn grunt dev"`
    - Before: `"build": "DEV=1 node --max_old_space_size=4096 ./node_modules/.bin/grunt build"`
    - After: `"build": "DEV=1 yarn grunt build"`


## Testing

1. JSX hot replacement:
    - Load: http://localhost-studio.code.org:9000/s/allthethings/lessons/18/levels/20
    - modify `GameButtons.jsx`, change `<div>{props.runButtonText || msg.runProgram()}</div>` to `<div>FunButton</div>`
    - Verify [Run] button changes to [Fun Button] without a reload
    - Verify watch mode works too, load: http://localhost-studio.code.org:3000/s/allthethings/lessons/18/levels/20 (note: port 3000), and verify that [Run] button has changed to [Fun Button] there too
2. TSX hot replacement:
    - Load: http://localhost-studio.code.org:9000/s/dance-ai-2023/lessons/1/levels/3
    - modify `SongSelector.tsx`, change `fa-play-pause` to `fa-play`
    - Verify the play/pause icon changes to a play icon without a reload
3. Storybook hot replacement:
    - Run `yarn storybook`, and open the Checkbox design component
    - Edit `Checkbox.tsx` and add text right before the `<input`
    - Verify that text appears in storybook without a reload
5. Verify `yarn build:dist` works and does NOT include HMR hooks
6. Verify `yarn build` works and does NOT include HMR hooks

Co-authored-by: Ben Brooks <ben@code.org>
